src/sentry/snuba/metrics/extraction.py

```diff
@@ -180,69 +215,171 @@ def query_hash(self) -> str:
+        assert tokens, "This query should not use on demand metrics"
+        return SearchQueryConverter(tokens).convert()
 
-        where = [c for c in (_convert_condition(c) for c in where) if c is not None]
 
-        if len(where) == 1:
-            return where[0]
-        else:
-            return {"op": "and", "inner": where}
+def _extract_field_info(aggregate: str) -> Tuple[Optional[str], str, MetricOperationType]:
+    """
+    Extracts the field name, metric type and metric operation from a Discover
+    function call.
 
+    This does not support derived metrics such as ``apdex`` and aggregates with
+    filters (``count_if``).
+    """
+    name, arguments, _alias = fields.parse_function(aggregate)
 
-def _extract_field_info(aggregate: str) -> Tuple[Optional[str], str, MetricOperationType]:
-    select = _get_query_builder().resolve_column(aggregate, False)
+    # TODO: Add support for derived metrics: failure_rate, apdex, eps, epm, tps, tpm
 
-    metric_type = _AGGREGATE_TO_METRIC_TYPE.get(select.function)
-    metric_op = _SNUBA_TO_METRIC_AGGREGATES.get(select.function)
-    assert metric_type and metric_op, f"Unsupported aggregate function {select.function}"
+    metric_type = _AGGREGATE_TO_METRIC_TYPE.get(name)
+    metric_op = _SEARCH_TO_METRIC_AGGREGATES.get(name)
+    assert metric_type and metric_op, f"Unsupported aggregate function {name}"
 
     if metric_type == "c":
-        assert not select.parameters, "Count should not have parameters"
+        assert not arguments, "`count()` does not support arguments"
         return None, metric_type, metric_op
     else:
-        assert len(select.parameters) == 1, "Only one parameter is supported"
+        assert len(arguments) == 1, "Only one parameter is supported"
+        return _map_field_name(arguments[0]), metric_type, metric_op
+
+
+def _map_field_name(search_key: str) -> str:
+    """
+    Maps a the name of a field in a search query to the event protocol path.
+
+    Raises an exception if the field is not supported.
+    """
+    # Map known fields using a static mapping.
+    if field := _SEARCH_TO_PROTOCOL_FIELDS.get(search_key):
+        return field
+
+    # Measurements support generic access.
+    if search_key.startswith("measurements."):
+        return f"event.{search_key}"
+
+    # Run a schema-aware check for tags. Always use the resolver output,
+    # since it accounts for passing `tags[foo]` as key.
+    resolved = (resolve_column(Dataset.Transactions))(search_key)
+    if resolved.startswith("tags["):
+        return f"event.tags.{resolved[5:-1]}"
+
+    raise ValueError(f"Unsupported query field {search_key}")
+
+
+QueryOp = Literal["AND", "OR"]
+QueryToken = Union[SearchFilter, QueryOp, ParenExpression]
+T = TypeVar("T")
 
-        name = select.parameters[0].name
-        assert name in _SNUBA_TO_RELAY_FIELDS, f"Unsupported field {name}"
 
-        return _SNUBA_TO_RELAY_FIELDS[name], metric_type, metric_op
+class SearchQueryConverter:
```

Comment:
This parser is so good, really really nice job. For now inline conversion is also more than fine, maybe in the future we could implement our own custom intermediate repr, so that we can have an independent visitor for generating the JSON but for now it's more than enough.